File: dsa/sorting/2357_make_array_zero_by_subtracting_equal_amounts.java
package sorting;

class Solution {
    public int minimumOperations(int[] nums) {
        int result = 0;
        int[] countVal = new int[101];
        int maxVal = 0;
        int culSub = 0;

        for (int i = 0; i < nums.length; i++ ){
            countVal[nums[i]] += 1;
            if (nums[i] > maxVal){
                maxVal = nums[i];
            }
        }
        while (maxVal > 0) {
            for (int i = 0; i < countVal.length; i++){
                if (countVal[i] > 0 && i > culSub){
                    int realSubVal = i - culSub;
                    maxVal = maxVal - realSubVal;
                    culSub += i - realSubVal;
                    countVal[i] -= 1;
                    result += 1;
                }
            }
        }
        return result;
    }
}

File: dsa/4.sorting/2357_Make_Array_Zero_by_Subtracting_Equal_Amounts.py
import typing as t

class Solution:
    def minimumOperations(self, nums: t.List[int]) -> int:
        freq_tracker: t.List[int] = [0] * 101
        acc_sum: int = 0
        max_val: int = 0
        result: int = 0
        
        for ind in range(0, len(nums), 1):
            freq_tracker[nums[ind]] += 1
            if nums[ind] > max_val:
                max_val = nums[ind]
        
        for ind in range(0, len(freq_tracker), 1):
            if freq_tracker[ind] > 0:
                curr_num: int = ind - acc_sum
                if curr_num > 0:
                    acc_sum += curr_num
                    max_val -= curr_num
                    result += 1
                
                if max_val == 0:
                    break        
        
        return result
        
        